locations/storefinders/amrest_eu.py

```diff
@@ -1,93 +1,150 @@
 from scrapy import Spider
-from scrapy.exceptions import CloseSpider
 from scrapy.http import JsonRequest
 
+from locations.automatic_spider_generator import AutomaticSpiderGenerator, DetectionRequestRule
 from locations.categories import Extras, apply_yes_no
 from locations.dict_parser import DictParser
 from locations.hours import OpeningHours
+from locations.items import Feature
 
 
-class AmrestEUSpider(Spider):
-    base_urls = ["https://api.amrest.eu/amdv/ordering-api/{}/"]
-    auth_url = "rest/v1/auth/get-token"
-    restaurants_url = "rest/v2/restaurants/"
-    base_headers = {"source": "WEB"}
-    # following 'deviceUuid' is the default for BK and KFC. PH uses generated one, but seems to work
-    # with default one as well
-    auth_data = {
-        "deviceUuid": "FFFFFFFF-FFFF-FFFF-FFFF-FFFFFFFFFFFF",
-        "deviceUuidSource": "FINGERPRINT",
-        "source": "WEB_KFC",
-    }
-
-    CHANNELS = {
-        "takeout": "TAKEOUT",
-        "takeaway": "TAKEAWAY",
-        "dineIn": "DINE_IN",
-        "delivery": "DELIVERY",
-    }
-
-    # Restaurants details endpoint url is different for different brands - for PH and BK involves channel.
-    # Usually it works out with TAKEOUT, but if it's not supported by restaurant, different channel was
-    # noticed to work fine.
-    def make_details_url(self, root_url, restaurant):
-        match self.item_attributes["brand"]:
-            case "KFC":
-                return f"{root_url}{self.restaurants_url}details/{restaurant['id']}"
-            case "Pizza Hut" | "Burger King" | "La Tagliatella":
-                for channel_key, channel in self.CHANNELS.items():
-                    if restaurant.get(channel_key):
-                        return f"{root_url}{self.restaurants_url}{restaurant['id']}/{channel}"
-            case _:
-                raise CloseSpider(f"Brand '{self.item_attributes['brand']}' not supported")
+class AmrestEUSpider(Spider, AutomaticSpiderGenerator):
+    api_brand_key: str = None
+    api_brand_country_key: str = None
+    api_source: str = None
+    api_auth_source: str = None
+    api_channel: str = None
+    detection_rules = [
```

Comment:
Is it worth extracting some of these repeated regexs to a constant/similar so to it's more obvious the difference between the detection rules is headers/data?

(Wasn't immediately obvious on reading it thats where the differences were, so I could see myself accidentally refactoring to "remove duplicates" and breaking things)

Reply:
Another option is perhaps to use `urlparse` for parsing URLs so that rules can be (and generally should be) defined with `urlparse` type rules (e.g. `{"api_key" = "key}"` means the generated spider should have attribute `api_key` set to the URL attribute `key`.

Reply:
Moved to https://github.com/alltheplaces/alltheplaces/issues/10202